### PR TITLE
Fix weight backward for cudnn conv of large tensor

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This is the last PR for https://github.com/pytorch/pytorch/issues/22496

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 4fe65747:


* **2/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=4fe65747305f6d324f43d3813f95f12f95fe1fad).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4141235) pytorch_linux_xenial_cuda9_cudnn7_py3_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=80786037) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=80786037))
<details>
<summary>
<code>Jan 06 19:15:47 /var/lib/jenkins/workspace/aten/src/ATen/native/cudnn/Conv.cpp:1191:150: error: 'raw_cudnn_convolution_backward_weight_out_32bit' was not declared in this scope</code>
</summary>

```
Jan 06 19:15:47                                    ^ 
Jan 06 19:15:47 /var/lib/jenkins/workspace/aten/src/ATen/core/interned_strings.h:214:3: note: in expansion of macro 'FORALL_ATTR_BASE_SYMBOLS' 
Jan 06 19:15:47    FORALL_ATTR_BASE_SYMBOLS(_)        \ 
Jan 06 19:15:47    ^ 
Jan 06 19:15:47 /var/lib/jenkins/workspace/aten/src/ATen/core/interned_strings.h:386:1: note: in expansion of macro 'FORALL_NS_SYMBOLS' 
Jan 06 19:15:47  FORALL_NS_SYMBOLS(DEFINE_SYMBOL) 
Jan 06 19:15:47  ^ 
Jan 06 19:15:47 /var/lib/jenkins/workspace/aten/src/ATen/native/cudnn/Conv.cpp:1172:145: error: 'raw_cudnn_convolution_backward_weight_out_32bit' was not declared in this scope 
Jan 06 19:15:47      raw_cudnn_convolution_backward_weight_out_32bit(grad_weight, grad_output, input, padding, stride, dilation, groups, benchmark, deterministic); 
Jan 06 19:15:47                                                                                                                                                  ^ 
Jan 06 19:15:47 /var/lib/jenkins/workspace/aten/src/ATen/native/cudnn/Conv.cpp:1191:150: error: 'raw_cudnn_convolution_backward_weight_out_32bit' was not declared in this scope 
Jan 06 19:15:47        raw_cudnn_convolution_backward_weight_out_32bit(grad_weight_, grad_output_, input_, padding, stride, dilation, groups, benchmark, deterministic); 
Jan 06 19:15:47                                                                                                                                                       ^ 
Jan 06 19:15:47 make[2]: *** [caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cudnn/Conv.cpp.o] Error 1 
Jan 06 19:15:47 make[2]: *** Waiting for unfinished jobs.... 
Jan 06 19:15:47 caffe2/CMakeFiles/torch_cuda.dir/build.make:2243: recipe for target 'caffe2/CMakeFiles/torch_cuda.dir/__/aten/src/ATen/native/cudnn/Conv.cpp.o' failed 
Jan 06 19:15:47 CMakeFiles/Makefile2:7285: recipe for target 'caffe2/CMakeFiles/torch_cuda.dir/all' failed 
Jan 06 19:15:47 make[1]: *** [caffe2/CMakeFiles/torch_cuda.dir/all] Error 2 
Jan 06 19:15:47 Makefile:140: recipe for target 'all' failed 
Jan 06 19:15:47 make: *** [all] Error 2 
Jan 06 19:15:47 -- Building version 1.4.0a0+4fe6574 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: test/test_nn.py [hunk 1]
```diff
@@ -9340,14 +9341,32 @@ def test_conv_transposed_large(self, device):
...
-        self.assertEqual(ret[:2048], conv1(input_large[:2048]))
-        self.assertEqual(ret[2048:], conv1(input_large[2048:]))
+        # forward
+        ret = conv(input_large)
+        self.assertEqual(ret[:2048], conv(input_large[:2048]))
+        self.assertEqual(ret[2048:], conv(input_large[2048:]))
+
+        # backward
+        conv.zero_grad()
+        ret.view(4096, -1).max(dim=1).values.sum().backward()
```
Comment (ngimel) [inline test/test_nn.py]: what's wrong with a simple sum() here?

File: test/test_nn.py [hunk 2]
```diff
@@ -9340,14 +9341,32 @@ def test_conv_transposed_large(self, device):
...
     def test_conv_large(self, device):
         dtype = torch.half if self.device_type == 'cuda' else torch.float
-        conv1 = nn.Conv2d(2, 2, 8, 8).to(device).to(dtype)
+        conv = nn.Conv2d(2, 2, 8, 8, bias=False).to(device).to(dtype)
         input_large = torch.randn(4096, 2, 512, 512, dtype=dtype, device=device)
-        ret = conv1(input_large)
-        self.assertEqual(ret[:2048], conv1(input_large[:2048]))
-        self.assertEqual(ret[2048:], conv1(input_large[2048:]))
+        # forward
+        ret = conv(input_large)
```
Comment (ngimel) [inline test/test_nn.py]: I suspect this is splitting the input into equal chunks so there's no tail - can you make it a test that has unequal split?

**Review by ngimel — COMMENTED** [review]
Generally looks good, small questions/suggestions about a test.

File: test/test_nn.py (hunk 1 above)
Comment (zasdfgbnm) [reply test/test_nn.py]: a simple `sum()` means all the outputs have contributions, therefore in the backward, we are summing lots of values, the rounding errors could then become as large as `1e-5`, larger than the threshold of `assertEqual` (`1e-6` I guess). Having a `.max(dim=1)` makes it sparse to reduce the rounding errors.

File: test/test_nn.py (hunk 2 above)
Comment (zasdfgbnm) [reply test/test_nn.py]: Sure I will

File: test/test_nn.py (hunk 1 above)
Comment (ngimel) [reply test/test_nn.py]: Cool, can you then please leave a comment about this?

Comment (zasdfgbnm) [reply test/test_nn.py]: Sure

File: test/test_nn.py (hunk 2 above)
Comment (zasdfgbnm) [reply test/test_nn.py]: I changed 4096 to 4097. Waiting for the CI status.

File: test/test_nn.py (hunk 1 above)
Comment (zasdfgbnm) [reply test/test_nn.py]: Comment added

**Review by ngimel — APPROVED** [review]
